From 670af4255632b413ea1683799ae392da4957fe7d Mon Sep 17 00:00:00 2001
From: Eric Richter <erichte@linux.ibm.com>
Date: Wed, 8 Jan 2025 11:59:24 -0600
Subject: [PATCH 2/2] tpm_i2c_nuvoton: Add support for Nuvoton 75x TPMs

This patch adds the new expected values for the 75x chip to the hdat i2c
devices table, and the requisite new constants to the Nuvoton driver as
according to the TCG TPM I2C Interfact Specification for TPM 2.0
Revision 1.0[1].

[1] https://trustedcomputinggroup.org/resource/tcg-tpm-i2c-interface-specification/

Signed-off-by: Eric Richter <erichte@linux.ibm.com>
Signed-off-by: Reza Arbab <arbab@linux.ibm.com>
---
 hdata/i2c.c                      |  1 +
 libstb/drivers/tpm_i2c_nuvoton.c | 16 +++++++++++++++-
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/hdata/i2c.c b/hdata/i2c.c
index 7d5d655a5354..235aae8a86bd 100644
--- a/hdata/i2c.c
+++ b/hdata/i2c.c
@@ -156,6 +156,7 @@ static struct hdat_i2c_type hdat_i2c_devs[] = {
 	{ 0x11, "eeprom", "atmel,24c16" },
 	{ 0x12, "i2c", NULL },   /* NVDIA GPU */
 	{ 0x13, "i2c", "nxp,lpc11u35" },
+	{ 0x15, "tpm", "nuvoton,npct75x" },
 };
 
 struct hdat_i2c_info {
diff --git a/libstb/drivers/tpm_i2c_nuvoton.c b/libstb/drivers/tpm_i2c_nuvoton.c
index e7cb477b6b13..70e34272fa92 100644
--- a/libstb/drivers/tpm_i2c_nuvoton.c
+++ b/libstb/drivers/tpm_i2c_nuvoton.c
@@ -49,6 +49,16 @@ static const struct tpm_info tpm_nuvoton_650 = {
 	.vid_did = 0x60,
 };
 
+static const struct tpm_info tpm_nuvoton_75x = {
+	.compatible = "nuvoton,npct75x",
+	.vendor_id = 0x5010FC00,
+	.sts = 0x18,
+	.burst_count = 0x19,
+	.data_fifo_w = 0x24,
+	.data_fifo_r = 0x24,
+	.vid_did = 0x48,
+};
+
 static struct tpm_dev *tpm_device = NULL;
 static const struct tpm_info *tpm_info = NULL;
 
@@ -626,8 +636,11 @@ static void __tpm_i2c_nuvoton_probe(const struct tpm_info *info)
 		/*
 		 * Tweak for linux. It doesn't have a driver compatible
 		 * with "nuvoton,npct650"
+		 *
+		 * Not necessary for 75x, as we use the compatible that
+		 * Linux expects.
 		 */
-		if (!dt_node_is_compatible(node, "nuvoton,npct601")) {
+		if (dt_node_is_compatible(node, "nuvoton,npct650")) {
 			dt_check_del_prop(node, "compatible");
 			dt_add_property_strings(node, "compatible",
 						"nuvoton,npct650", "nuvoton,npct601");
@@ -643,4 +656,5 @@ disable:
 void tpm_i2c_nuvoton_probe(void)
 {
 	__tpm_i2c_nuvoton_probe(&tpm_nuvoton_650);
+	__tpm_i2c_nuvoton_probe(&tpm_nuvoton_75x);
 }
-- 
2.43.5

